src/compiler/ruby_generator.cc

```diff
@@ -37,16 +38,61 @@ using std::vector;
 namespace grpc_ruby_generator {
 namespace {
 
+// returns the full canonical message name
+grpc::string GetCanonicalMessageType(const Descriptor* msg) {
+  const Descriptor* top_level_msg;
+  const FileDescriptor* file_containing_msg;
+  grpc::string msg_full_name;
+  grpc::string msg_name;
+  grpc::string msg_proto_pkg;
+  grpc::string resolved_namespace;
+  grpc::string res(msg->full_name());
+
+  // If msg is nested, find the topmost message
+  top_level_msg = msg;
+  while (top_level_msg->containing_type()) {
+    top_level_msg = top_level_msg->containing_type();
+  }
+  file_containing_msg = top_level_msg->file();
+
+  msg_full_name = top_level_msg->full_name();
```

Comment:
nit: can we declare and initialize these three vars down to just above where they're used?